active/0000-app-import-output-file-for-addons.md

```diff
@@ -0,0 +1,116 @@
+
+`outputFile:` option, specifies the target files for the given addon import. If multiple addons share an outputFile, they will be concatenated in the order they where imported, but appended to regular `app.import`s.
+
+`outputFile:` will default to `assets/vendor.js` and `assets/vendor.css`
+
+Today all addons defined in package.json are processed and imported. This behavior remains the same. `app.import` is used for consistency with how we import `vendor` files, but it is optional for addons and only needed if we want to override the outputFile.
+
+While today we can't `app.import` from `node_modules`, this might be supported in the future and other consumers might already be funneling to make this work, in order to dissambiguate we import them from `addons` instead as `app.import('addons/ember-view-state', ...)`. This is also more explicit since we're importing the processed JS, instead of a specific file under `node_modules`.
+
+## Examples
+
+
+#### variation 0: the default
+
+```js
+app.import('addons/ember-lgtm', { outputFile: 'assets/vendor.js'});
+```
+
+The code above is redundant, since it is the default for all addons. This is only needed if you need to specify the order or override the `outputFile`.
+
+#### variation 1: 1 file -> 1 outputFile
+
+```js
+app.import('addons/ember-lgtm', { outputFile: 'assets/ember-lgtm.js'});
+```
+
+* The treeForAddon for  `ember-lgtm` becomes `assets/ember-lgtml.js`
+* Other addon trees
+* in prod it is:
+  * uglified (unless using the uglify options it is excluded)
+  * fingerprinted (unless it is excluded via the asset-rev options)
+
+#### variation 2, multiple files to same outputFile
+
+```js
+app.import('addons/ember-signature-pad.js', { outputFile: 'assets/vendor-deferred.js'});
+app.import('vendor/background-update.js', { outputFile: 'assets/vendor-deferred.js'});
+```
+
+* in-order of the corresponding `app.import` invocation, using sourceMap
+  concat, the files are combined into `assets/vendor-deferred.js`
+
+#### variation n, multiple files to same outputFile
+
+```js
+app.import('addons/ember-addon-1.js', { outputFile: 'assets/alternate-vendor.js'});
+app.import('addons/ember-addon-2.js', { outputFile: 'assets/alternate-vendor.js'});
+app.import('addons/ember-addon-n.js', { outputFile: 'assets/alternate-vendor.js'});
+```
+
+# Drawbacks
+
+* Potential confusion about static and "addon" files. It isn't clear which ones are procesed and which ones are not.
+* Potential confusion about using `addons/` as the path.
+* Potential confusion about the order since addons not specified are still "imported" into vendor.js and vendor.css
```

Comment:
> which parts of the app depend on which add-ons will prove to be confusing, how will this be addressed?

I don't think this is any different from what we have for static vendor assets. I think a proper dep-graph could certainly help, but I'm not sure it is riskier than what we can already do today. 

> what happens if an add-on has a app directory, do we detect that and error if they try to app.import ?
 
I think that throwing would be simpler. However, if the addon's app.import type is for `vendor` we can add it to the same `outputFile`, and maintain the current behavior for type `test`. The problem with adding it to outputFile is when we have more than one addon that imports the same file. Today we'll add the content only once, but if the outputFile is different I think we should at least warn that both `outputFile`s will contain the same content.